### ISSUE TITLE
Design to allow import to database from cli

### ISSUE THREAD

**Issue by 1dotd4** [1dotd4/go#2]
It should be defined an API that allows CRUD operations on the data. This allow to integrate more scripts and more methods of feeding data to the overview.

**Comment by 1dotd4** [comment]
This is a program design problem.
Should I run the server and then use import to import to the server so that it could be done with some cron or other methods like webhooks?
Or should it be executed in separate times for importing data to the database and then served?
Let's see more use cases.

**Comment by 1dotd4** [comment]
Note that the only API I would need are those for inserting data or asking for a specific view of the database.

**Comment by 1dotd4** [comment]
The following is a sample of use case
```sh
cd /path/to/workdir
for repo in "list of repository names"; do
  git clone git@workrepo:$repo $repo
  git-overview --import ./$repo
done
git-overview --serve &
```

This leads to the question, should be the --import function to import the data or the scheduled update inside --serve to import it?
As this is supposed to be an all in one binary, I would do the actual import before starting the service; a parallel task will fetch and update logs periodically.
Continuing in this idea we have the following components:
- http server for redering pages;
- http server for webhooks (on a different path, like `/webhook/`, TBD);
- daemon that fetch updates from git directories;
- (most importantly) the core interface that manage the database. 

SQLite3 is already [thread safe](https://www.sqlite.org/threadsafe.html) which allow us to write a simple interface for our database that can be used by the other three components.

To recap, --import should just add the repository path to a "will do" list that will completed at startup of the service. It will be saved the last synchronized commit for each repository in the database to instruct git to log only the last one. 

Note: _Reading while writing should be tested._

**Comment by 1dotd4** [comment]
Meanwhile the web hook should work as follow:
```
$ git-overview --add-webhook github a-repository-name
Added repository a-repository-name with web hook token <token-here>
Use `/webhook/<token-here>` in GitHub.
```

So there should be stored in the database if the repository is either a local git folder or is a remote repository contacting the server. Possible cloud SCM are GitHub, Gitea, GitLab and should be easy to implement a new web hook systems.

**Comment by 1dotd4** [comment]
About threads:

> Threads are evil. Avoid them.
>
> SQLite is threadsafe. We make this concession since many users choose to ignore the advice given in the previous paragraph. [...]

[SQLite FAQ](https://www.sqlite.org/faq.html#q6)
[Parallel read and write in SQLite](https://www.skoumal.com/en/parallel-read-and-write-in-sqlite/)

So in case the program hangs to serve a query, one may use [WAL](https://www.sqlite.org/wal.html) to make writes bulk.

**Comment by 1dotd4** [comment]
[GitHub webhooks](https://docs.github.com/en/developers/webhooks-and-events/webhooks/webhook-events-and-payloads)
[GitLab webhooks](https://docs.gitlab.com/ee/user/project/integrations/webhooks.html)
[Gitea webhooks](https://docs.gitea.io/en-us/webhooks/)

**Comment by 1dotd4** [comment]
`git --no-pager log --format='format:"%H","%at","%an","%ae","%s"'`

**Comment by 1dotd4** [comment]
`git --no-pager log --format='format:%H%x09%at%x09%an%x09%ae%x09%s'`

**Comment by 1dotd4** [comment]
Should I manage the entire tree?
This simplifies the false fact that each commit has a branch and allow external join on the tree to concurrent commits in two branches.

**Comment by 1dotd4** [comment]
```scheme
(car
  (with-input-from-pipe
    "git --git-dir=./test/a-repo/.git --no-pager log --branches --tags --remotes --full-history --date-order --format='format:%H%x09%P%x09%at%x09%an%x09%ae%x09%s%x09%D'"
     (lambda ()
      (map
        (lambda (a) 
                  (string-split a "\t" #t))
        (read-lines)))))
```

**Comment by 1dotd4** [comment]
Currently the concern is in two data design problems:

(1) I need to see in which branch the last commit made by someone was.
To do this I think I need to check with a recursive table if the commit is in a branch.
`git branch -v --no-abbrev` gives me the head of the branch.
A problem with this is distinguishing whether a branch was already merged to master or not.
In other words recreating the graph of git.

(2) Select commits of a specific branch.
This is similar as I need to know where the branch started. Which does not look trivial right now.

**Comment by 1dotd4** [comment]
Maybe scanning the main branch and then doing an exclusion while descending the tree.
Using the fact that I know where they merge as I have a primary parent that I can select it's plausible.

So the "flow" of the query looks like:
1. select the head of selected branches;
2. list the commits descending the branches;
3. if not wanted, exclude the main branch.

This adds a table for CommitsParents

**Comment by 1dotd4** [comment]
To recap, now we have a database for commits that works for both cli. Webhooks are gonig to be available in another release, probably there will be adapters for OAuth as seen in #5 so I don't have to think about it now.
What am I missing before wrapping up?

**Comment by 1dotd4** [comment]
Added `-r` in `git branch` so now I'm working with just the stuff that's on the scm server. So far this issue looks resolved.
